File: src/main/java/com/learning/leetcode/arrays/medium/PairOfSongs.java
package com.learning.leetcode.arrays.medium;

import java.util.Arrays;

public class PairOfSongs {
    public static void main(String[] args) {
        int[] time = {30, 20, 150, 100, 40};
        System.out.println("Given Array : " + Arrays.toString(time));
        System.out.println("No of pairs divisible by 60 : " + numPairsDivisibleBy60(time));
        time = new int[]{60, 60, 60};
        System.out.println("Given Array : " + Arrays.toString(time));
        System.out.println("No of pairs divisible by 60 : " + numPairsDivisibleBy60(time));

    }

    private static int numPairsDivisibleBy60(int[] time) {
        int[] moduloArr = new int[60];
        int count = 0;
        for (int t : time) {
            if (t % 60 == 0) {
                count += moduloArr[0];
            } else {
                count += moduloArr[60 - t % 60];
            }
            moduloArr[t % 60]++;
        }
        return count;
    }
}

File: src/main/java/com/learning/leetcode/arrays/medium/PairOfSongs.py
from typing import List


def numPairsDivisibleBy60(time: List[int]) -> int:
    arr = [0] * 60
    count = 0
    for t in time:
        if t % 60 == 0:
            count += arr[0]
        else:
            count += arr[60 - t % 60]
        arr[t % 60] += 1
    return count


def main():
    time = [30, 20, 150, 100, 40]
    print("Given Array : ", time)
    print("No of pairs divisible by 60 : ", numPairsDivisibleBy60(time))
    time = [60, 60, 60]
    print("Given Array : ", time)
    print("No of pairs divisible by 60 : ", numPairsDivisibleBy60(time))
    time = [60, 60, 60, 60]
    print("Given Array : ", time)
    print("No of pairs divisible by 60 : ", numPairsDivisibleBy60(time))


main()
